[ovirt] dynamic inventory plugin

bverschueren opened the pull request:
##### SUMMARY
This PR introduces a dynamic inventory plugin for oVirt.

##### ISSUE TYPE

- Feature Pull Request

bverschueren committed:
add ovirt inventory plugin

fix faulty NonType mapping and pep violation

add interface selection option

Implement caching & only return ansible_host if at least one device is returned

return list of vm instead of dict

bverschueren commented:
cc @machacekondra

Akasurde commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+# -*- coding: utf-8 -*-
+#
```
Replace line 2 - 19 with 

```
# Copyright: (c) 2016, Red Hat, Inc.
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

bverschueren replied:
fixed this

Akasurde commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+      - Uses 'ovirt.yml', 'ovirt4.yml', 'ovirt.yaml', 'ovirt4.yaml' YAML configuration file.
+    options:
+      plugin:
+        description: the name of this plugin, it should alwys be set to 'ovirt' for this plugin to recognize it as it's own.
```
```suggestion
        description: the name of this plugin, it should always be set to 'ovirt' for this plugin to recognise it as it's own.
```

Akasurde commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+        required: True
+        choices: ['ovirt']
+      ovirt_url:
+        description: url to ovirt-engine api
```
```suggestion
        description: URL to ovirt-engine API.
```

Akasurde commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+        description: url to ovirt-engine api
+        required: True
+      ovirt_username:
+        description: ovirt authentication user
```
```suggestion
        description: ovirt authentication user.
```

Akasurde commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+        description: ovirt authentication user
+        required: True
+      ovirt_password:
+        description: ovirt authentication password
```
```suggestion
        description: ovirt authentication password.
```

Akasurde commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+        description: ovirt authentication password
+        required : True
+      ovirt_cafile:
+        description: path to ovirt-engine ca file
```
```suggestion
        description: path to ovirt-engine CA file.
```

Akasurde commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+        description: path to ovirt-engine ca file
+        required: False
+      ovirt_preferred_interface:
+        description: vm network interface from which the ansible_host ip is selected
```
```suggestion
        description: vm network interface from which the ansible_host IP is selected.
```

Akasurde commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+'''
+
+EXAMPLES = '''
+# ovirt.yml
```
```suggestion
# Sample content of ovirt.yml
```

Akasurde commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+try:
+    import ovirtsdk4 as sdk
+except ImportError:
+    print('oVirt inventory script requires ovirt-engine-sdk-python >= 4.2.4')
```
I would raise `AnsibleError` like 

```suggestion
    HAS_OVIRT_LIB = False
    ...
    # in def verify_file
    if not HAS_OVIRT_LIB:
        raise AnsibleError('oVirt inventory script requires ovirt-engine-sdk-python >= 4.2.4')
```

bverschueren replied:
fixed this

bcoca commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+        - inventory_cache
+    description:
+      - Get inventory hosts from the ovirt service.
+      - Uses 'ovirt.yml', 'ovirt4.yml', 'ovirt.yaml', 'ovirt4.yaml' YAML configuration file.
```
s/Uses/Requires a file ending in/

bverschueren replied:
fixed this

bcoca commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,294 @@
+      ovirt_cafile:
+        description: path to ovirt-engine ca file
+        required: False
+      ovirt_preferred_interface:
```
instead of requiring specific options to configure specific host variables, inherit from constructed and allow the user to create his own expressions.

 You assume that the devices are configured following a pattern, while it should be that way in many cases, it isn't always. Users might want to match `ansible_host` to a specific ip range, cluster/iface/tags combos, etc. The compose engine allows much more flexibility in this case.

bverschueren replied:
I've updated the plugin to inherit from constructed and allow more flexible expressions. Also improved the vm filtering to be more in the user's control.

Akasurde committed:
Apply suggestions from code review

Co-Authored-By: bverschueren <bverschueren@users.noreply.github.com>

bverschueren committed:
improve search/grouping to support more flexible expressions

bverschueren committed:
fix pylint violation

AlanCoding commented:
👋 just wanted to drop in and say hi, because the `ovirt4.py` script is used in inventory updates in [AWX](https://github.com/ansible/awx) and Ansible Tower. We will transition to inventory plugins (like this one) when they become available. @kdelee and I would like to help test and review this.

Do you think this will target going into Ansible 2.9?

Also, at first look, it appears that PR https://github.com/ansible/ansible/pull/54338 is relevant to this. That's the only really major change to the script I see since this PR was opened.

willtome commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+        description: the name of this plugin, it should always be set to 'ovirt' for this plugin to recognise it as it's own.
+        required: True
+        choices: ['ovirt']
+      ovirt_url:
```
Can this also support using the `OVIRT_URL` environment variable used by ovirt modules?
```
env:
  - name: OVIRT_URL
```

willtome commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+      ovirt_url:
+        description: URL to ovirt-engine API.
+        required: True
+      ovirt_username:
```
Can this also support using the `OVIRT_USERNAME` environment variable used by ovirt modules?
```
env:
  - name: OVIRT_USERNAME
```

willtome commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+      ovirt_username:
+        description: ovirt authentication user.
+        required: True
+      ovirt_password:
```
Can this also support using the `OVIRT_PASSWORD` environment variable used by ovirt modules?
```
env:
  - name: OVIRT_PASSWORD
```

willtome reviewed (commented):
Add support for environment variables as authentication parameters. Is there an update on getting this merged?

s-hertel commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+
+        query_filter = self._get_query_options(self.get_option('ovirt_query_filter', None))
+
+        if cache:
```
The value of `cache` here reflects whether or not `- meta: refresh_inventory` has been used. And even if it's False (i.e. the cache needs to be refreshed) if the user has enabled caching (i.e. the value of self.get_option('cache')) then the cache should be updated. Here's some documentation that provides a pattern to follow: https://docs.ansible.com/ansible/latest/dev_guide/developing_inventory.html#inventory-cache

bverschueren replied:
I've updated the caching mechanism in line with how it's described in the link. The documentation is much clearer now then it was a year ago.
However, I think the doc misses the case where a user has not enabled caching. In that case get_inventory() is never called.

s-hertel commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+
+        if cache:
+            try:
+                source_data = self.cache.get(cache_key)
```
`self.cache.get(cache_key)` and `self.cache.set(cache_key, source_data)` shouldn't be used for caching because it's a deprecated interface. `self._cache` should be used instead as a dictionary. The documentation linked above has examples.

s-hertel commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+
+    def verify_file(self, path):
+
+        if not HAS_OVIRT_LIB:
```
Move this check to `parse`. If an exception is raised here and the plugin is in the inventory whitelist the plugin won't be used to parse the source and the user will not see any indication as to why.

s-hertel commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+
+DOCUMENTATION = '''
+    name: ovirt
+    plugin_type: inventory
```
The documentation could use a `version_added` and `author` field.

s-hertel commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+          :param host: dict representation of oVirt VmStruct
+          :param return: preferred hostname for the host
+        '''
+        hostname_preference = self.get_option('ovirt_hostname_preference', ['fqdn, name'])
```
The second argument of get_option is actually hostvars, so this doesn't do what you think. Since `['fqdn', 'name']` is the default for the ovirt_hostname_preference option anyway, this should just be:
```
hostname_preference = self.get_option('ovirt_hostname_preference')
```
You may want to check that it's truthy before progressing, and raise an AnsibleParserError if the user has provided an empty list or None.

s-hertel commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+
+    def _get_hostname(self, host):
+        '''
+          Get the host's hostname based on prefered attribute
```
```suggestion
          Get the host's hostname based on preferred attribute
```

mnecas replied:
@bverschueren please could you correct this typo

s-hertel commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+            elif preference == 'name':
+                hostname = host.get('name')
+            else:
+                raise AnsibleParserError("%s is not a valid hostname precedent" % preference)
```
If those are the only two valid choices, also add `choices: ['fqdn', 'name']` to the documentation for the option. Otherwise, perhaps this section should just be
```
        for preference in hostname_preference:
            hostname = host.get('preference')
            if hostname is not None:
                return hostname
        raise AnsibleParserError("No valid name found for host id={0}".format(host.get('id')))
```

s-hertel commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+        else:
+            return vms_service.list()
+
+    def _add_hosts(self, hosts, group):
```
This method appears to be unused and can be removed.

s-hertel commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,282 @@
+            except KeyError:
+                update_cache = True
+            else:
+                self._populate_from_source(source_data)
```
`_populate_from_source` is occurring twice if the cache is used.

s-hertel reviewed (changes requested):
This plugin needs integration tests to ensure backwards compatibility with the ovirt contrib script (it must be possible for the plugin to create the same groups/hosts/vars). Here's an example of a test created for the AWS inventory plugin and script: https://github.com/ansible/ansible/tree/devel/test/integration/targets/inventory_aws_conformance and here's one in progress for the GCP inventory plugin and script: https://github.com/ansible/ansible/pull/54701

bverschueren committed:
several plugin improvements

 - move sdk version check to parse function
 - add version_added and author documentation fields
 - allow any list as hostname_preference
 - support ovirt credentials from env
 - remove unused add_hosts method
 - align cache mechanism with standard pattern

bverschueren committed:
add integration tests

willtome commented:
Should this be shipped with a collection containing the other Ovirt roles (https://galaxy.ansible.com/oVirt) to align with the future Ansible content strategy?

bverschueren committed:
use env variables for the inventory script

Configparser in python 3.7+ does not allow empty config values by
default and thus we use env variables in the test if we don't want to
change the script.

bverschueren commented:
ready_for_review

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+
+HAS_OVIRT_LIB = False
+
+try:
+    import ovirtsdk4 as sdk
+    HAS_OVIRT_LIB = True
+except ImportError:
+    HAS_OVIRT_LIB = False
```
Why cannot you use here `AnsibleError('oVirt inventory script requires ovirt-engine-sdk-python >= 4.2.4')`?

bverschueren replied:
It was suggested (https://github.com/ansible/ansible/pull/49085#discussion_r340228609) to move this to the parse() function.

mnecas replied:
Thank you for information.

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+                url=self.ovirt_engine_url,
+                username=self.ovirt_engine_user,
+                password=self.ovirt_engine_password,
+                ca_file=self.ovirt_engine_cafile
```
Add: `insecure=not bool(self.ovirt_engine_cafile),`

bverschueren replied:
Shall we simply assume (but document) that the absence of a CA file leads to an insecure connection ? Or shall I introduce an `insecure` option to control this ?

mnecas replied:
you can add the option `insecure` but default would be as `not bool(self.ovirt_engine_cafile)`  as it is in https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/ovirt/ovirt_auth.py

bverschueren replied:
Right, I'll add an `insecure` option, but, imo it's rather dangerous that not specifying a CA automatically leads to an insecure connection (as is the case with ovirt_auth). 
I'd rather :
1) use provided CA (ignore `insecure`) or,
2) use system-wide if no CA provided (given `insecure` defaults to False)
3) a) fail (given `insecure` defaults to False)
    b) continue insecurely if user specifically sets `insecure=True`

whereas ovirt_auth:
1) use provided CA (ignore `insecure`) or,
2) a) continue insecurely if no CA is provided (`insecure` defaults to None, toggling to True if no CA is provided)
https://github.com/ansible/ansible/blob/0c597f4434d5dcde56d5b28ad62dc22bd0b4babf/lib/ansible/modules/cloud/ovirt/ovirt_auth.py#L261
b) fail if no CA is provided AND user specifically sets `insecure=False`

mnecas replied:
okay good idea

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+                if vm.name in [vm.name for vm in connection.follow_link(group.vms)]
+            ],
+            'statistics': dict(
+                (stat.name, stat.values[0].datum) for stat in stats
```
Had some issues with this in my env but this fixed it:
`(stat.name, stat.values[0].datum if stat.values else None) for stat in stats`

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+        connection = self._get_connection()
+
+        vms_service = connection.system_service().vms_service()
+        if query_filter is not None:
+            return vms_service.list(**query_filter)
+        else:
+            return vms_service.list()
```
Please replace with
```python 
        if query_filter is not None:
            return vms_service.list(**query_filter)
        return vms_service.list()
```

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+        description: path to ovirt-engine CA file.
+        required: False
+      ovirt_query_filter:
+        required: false
```
`required: False`

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+
+EXAMPLES = '''
+# Sample content of ovirt.yml
+# plugin: ovirt
+# ovirt_url: http://localhost/ovirt-engine/api
+# ovirt_username: ansible-tester
+# ovirt_password: secure
+# ovirt_query_filter:
+#   search: 'name=myvm AND cluster=mycluster'
+#   case_sensitive: no
+#   max: 15
+# keyed_groups:
+#   - key: cluster
+#     prefix: 'cluster'
+# groups:
+#   dev: "'dev' in tags"
+# compose:
+#   ansible_host: devices["eth0"][0]
```
Please remove the `# `.

mnecas commented:
Thank you once again @bverschueren for this PR.
After ansible devs will finish filters documentation for colletion I'll move the patch there.

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+
+        config = self._read_config_data(path)
+
+        self.ovirt_engine_url = self.get_option('ovirt_url')
+        self.ovirt_engine_user = self.get_option('ovirt_username')
+        self.ovirt_engine_password = self.get_option('ovirt_password')
+        self.ovirt_engine_cafile = self.get_option('ovirt_cafile')
```
Maybe create here the `self.connection` instead of in the `_get_connection` method.

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+
+        valid = False
+        if super(InventoryModule, self).verify_file(path):
+            if path.endswith(('ovirt.yml', 'ovirt4.yml', 'ovirt.yaml', 'ovirt4.yaml')):
```
I'm not sure with logic that the plugin would accept only files which ends with `ovirt.yml` etc.
What if it would accept any file?

willtome replied:
The convention across inventory plugins is that the file is in the format `*<NAME>.yml|yaml`. It is only really governed by this method but this is consistent with other plugins.

mnecas replied:
okay if its standard keep it this way

AlanCoding commented:
Want to make sure @jladdjr is aware of activity here.

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+        description: dictionary of filter key-values to query VM's. See U(https://ovirt.github.io/ovirt-engine-sdk/master/services.m.html#ovirtsdk4\
+.services.VmsService.list) for filter parameters.
+      ovirt_hostname_preference:
+        required: false
```
`required: False`

mnecas commented on lib/ansible/plugins/inventory/ovirt.py:
```diff
@@ -0,0 +1,275 @@
+#   search: 'name=myvm AND cluster=mycluster'
+#   case_sensitive: no
+#   max: 15
+# keyed_groups:
+#   - key: cluster
+#     prefix: 'cluster'
+# groups:
+#   dev: "'dev' in tags"
+# compose:
+#   ansible_host: devices["eth0"][0]
```
Just an idea but it would be great to explain how the `keyed_groups` and `groups` works and what it does (I might miss some docs but I have not found any good explanations)

bverschueren replied:
This is explained here: https://docs.ansible.com/ansible/latest/plugins/inventory.html#using-inventory-plugins

mnecas commented:
Please continue on this PR here https://github.com/oVirt/ovirt-ansible-collection. I have migrated all modules, module_utils etc. also this PR to our collection.

AlanCoding commented:
I think that what @mnecas did is the right thing to do. This main Ansible repo will be locked soon, if not already, and changes like these will no longer accepted. This should be closed, and future modifications should go to that collection. I appreciate you putting the inventory plugin in.

bverschueren commented:
This is followed-up in https://github.com/oVirt/ovirt-ansible-collection/pull/3
